Add a postprocessor for side integration of functors

GiudGiud opened the pull request:
In support of SAM-Pgh coupling
This is a draft right now. My awesomely complex test is failing, showing I dont understand face infos :(

GiudGiud committed:
Allow a postprocessor for the outlet pressure BC, refs #18086

GiudGiud committed:
Add a postprocessor for performing integrals of functors on surfaces, refs #18086

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/992279/) on 2b1d0ef wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20622/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20622/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7c8de46fce1c0b2302d905a0b7081604d7f7281c`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/992308/) on a4d15ec wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20622/site)

This comment will be updated on new commits.

lindsayad commented:
PIng me when you're ready for review

GiudGiud committed:
Add core of side integral of a functor postprocessor, refs #18086
Attempts at elem sidedness mindfulness

GiudGiud committed:
Remove unnecessary header from INSFVBodyForce

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit ec3954e32a5f6b4d71853a8c007896417aebce58
+Subproject commit 544e2c87ea2f35aab900dde8b8f1e0926d720421
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1006360/) on c00bee5 wanted to post the following:

The following commit(s) contain fixups:

- d655e8e606: fixup! Add a postprocessor for performing integrals of functors on surfaces, refs #18086

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1006360/) on c00bee5 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

GiudGiud committed:
Add block restriction hasBlocks API to functors.
This is a bit of a tough call. We d love to be able to use the BlockRestrictableInterface but this interface also does parameter work, which functors being so simple & agile do not dwell in. There s also a lot of other routines in that interface that simply dont apply to functors. So it seems this interface would at least need to be split. Even then this does not solve the problem of where it would be inherited. For example for variables it is inherited at MooseVariableBase, so very high in the chain, while functors only come in at MooseVariableField. We obviously cant inherit it twice in that tree.
So for now we provide a reduced BRI inside MooseFunctor, linking to the actual BRI in the classes that are both functors and inherit the BRI.

GiudGiud committed:
Move the work to get the FaceInfo in the SideUserObject since many classes should use it
Use proper face info when performing integration on mesh internal sidesets

Refs #18086

GiudGiud committed:
typo fix

GiudGiud committed:
Use better API for volumetric flow rates on internal sides
refs #18086

GiudGiud committed:
element from the sideset may not be the element that has a face info
Fix up tests to be more mindful of functor domain definition
Refs #18086

GiudGiud committed:
Fixup getter for face info to handle some of the refinement issues
Output an error when we are out of scope of what s possible in an element-side loop
refs #18086

GiudGiud committed:
Add tests for unsupported cases

Turn on refinement to find bugs

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1021006/) on c3e9460 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20622/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20622/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c697a44fc59c7cd9b24422f5191fb6c24f93d416`

GiudGiud commented:
@lindsayad this is ready for review.

there is one glaring issue here, which is that the sideUO loop isnt looping on face infos but on elements and sides
This doesnt work for interfaces where elements on both sides are not at the same level of refinement. The faceInfo should connect the active elements and the active more-refined neighbors. The current_elem_side / current_side etc from assembly point to the inactive one, and we cant just:
- pick a child at random (this would be the wrong centroid location for example)
- pick the inactive neighbor (this would work for functions, like the position would be right, but break for variables)

I did some clean up, made an as resilient as possible routine, and added a clear error message. 

This should help a lot measuring flow rates.

GiudGiud commented:
@cticenhour should be good now

GiudGiud committed:
Fix test failures

GiudGiud commented:
We should consider ditching the InternalVolumetricFlowRate for finite volume. The distinction between interfaceUO and sideUO doesnt make sense for functor-FV

lindsayad commented:
regarding adaptivity, have you looked over the `loopOverElemFaceInfo` function in `FVUtils.h`? I don't know whether there is some inspiration there for you. It may be that there is not

GiudGiud commented:
It is the solution, we should move SideUOs to a loop over face info

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1021845/) on 6e6d939 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>483d31</th>
      <th colspan=4>#20622 6e6d93</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.59%</td>
      <td>82.59%</td>
      <td>+0.00%</td>
      <td>83.33%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>74260</td>
      <td>74305</td>
      <td>+45</td>
      <td>45</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15657</td>
      <td>15665</td>
      <td>+8</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20622/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20622/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>483d31</th>
      <th colspan=4>#20622 6e6d93</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.68%</td>
      <td>83.67%</td>
      <td>-0.01%</td>
      <td>85.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>8419</td>
      <td>8428</td>
      <td>+9</td>
      <td>17</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1642</td>
      <td>1645</td>
      <td>+3</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20622/coverage_diff/navier_stokes.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20622/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20622/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 83.33% is less than the suggested 90.0%
- `navier_stokes` new line coverage rate 85.00% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented:
> It is the solution, we should move SideUOs to a loop over face info

What do you mean by this? Like that code has shown you a solution that you can implement easily? Or in order for that code to be relevant/useful SideUOs would have to only loops over face information, as opposed to elements and sides?

GiudGiud commented:
There are more FaceInfos than there are elements and sides (excluding duplicates, assuming you only loop once on each relevant face). I m actually not as clear as I should be on what the SideUO loop is, the SideUO loop is the one of the most obscure. It's an onBoundary case in the generalUO loop as far as I understand. 

That code you point me to works for flux kernels, so it does the right thing wrt to picking faces to iterate on.
Switching over to this loop is a larger rework than I want to do for this PR. Maybe it is easy, I have not tried.

lindsayad commented on framework/include/userobject/SideUserObject.h:
```diff
@@ -48,4 +52,47 @@ class SideUserObject : public UserObject,
+   * Note this face info could hold the element from the other side of the
+   * sideset. Sidesets are oriented!
+   */
+  const FaceInfo * getFaceInfo()
+  {
+
+    // Either the element or the neighbor is a valid argument to get a face info
+    const Elem * neighbor = _current_elem->neighbor_ptr(_current_side);
+    const Elem * element = _current_elem;
+    int side = _current_side;
+
+    mooseAssert(_current_elem, "We should have an element");
+    mooseAssert(_current_elem->active(), "The current element should be active");
+
+    // No neighbor means we are at a boundary, a FaceInfo exists in the mesh
+    // If a neighbor exists, the face info may only be defined on that side
+    if (neighbor)
+    {
+      // neighbor is not active and we cant just pick an arbitrary refined element in neighbor
+      // no FaceInfo in mesh will satisfy us, let's error out
+      if (!neighbor->active())
+        mooseError("Inactive neighbor found in SideUO loop. Uneven refinement across sidesets "
+                   "is not supported by the faceInfo system in user objects");
+
+      // First check refinement level
+      if (_current_elem->level() < neighbor->level())
+      {
+        element = neighbor;
+        side = neighbor->which_neighbor_am_i(_current_elem);
+      }
+      // Then check ids
+      else if (_current_elem->id() > neighbor->id())
+      {
+        element = neighbor;
+        side = neighbor->which_neighbor_am_i(_current_elem);
+      }
+    }
+
+    return _mesh.faceInfo(element, side);
+  }
```
```suggestion
void
getFaceInfo()
{
  _face_infos.clear();

  // Either the element or the neighbor is a valid argument to get a face info
  const Elem * equal_level_neighbor = _current_elem->neighbor_ptr(_current_side);
  _candidate_neighbors = {equal_level_neighbor};

  mooseAssert(_current_elem, "We should have an element");
  mooseAssert(_current_elem->active(), "The current element should be active");

  // No neighbor means we are at a boundary, a FaceInfo exists in the mesh
  // If a neighbor exists, the face info may only be defined on that side
  if (equal_level_neighbor)
  {
    // neighbor is not active and we cant just pick an arbitrary refined element in neighbor
    // no FaceInfo in mesh will satisfy us, let's error out
    if (!equal_level_neighbor->active())
      equal_level_neighbor->active_family_tree_by_neighbor(_candidate_neighbors, _current_elem);

    for (const Elem * neighbor : _candidate_neighbors)
    {
      const Elem * element = _current_elem;
      auto side = _current_side;

      // First check refinement level
      if (_current_elem->level() < neighbor->level())
      {
        element = neighbor;
        side = neighbor->which_neighbor_am_i(_current_elem);
      }
      // Then check ids
      else if (_current_elem->id() > neighbor->id())
      {
        element = neighbor;
        side = neighbor->which_neighbor_am_i(_current_elem);
      }
      _face_infos.push_back(_mesh.faceInfo(element, side));
    }
  }
  else
    _face_infos.push_back(_mesh.faceInfo(_current_elem, _current_side));
}
```

lindsayad replied:
Oopsie, I didn't update the comments, but I think the above should do what you need for mixed level refinement. This should also go in the .C file

GiudGiud replied:
this means that in every postprocessor you will have to nest a loop on the faceInfos. This is a little out of line with a loop on quadrature points imo

GiudGiud replied:
umm. Maybe I am being too narrow minded. Like the faceInfo loop will only be triggered for functors that work with FaceInfos. 
The functors that work with qp on the side (FE variables) will not need faceInfos

lindsayad replied:
That seems like a pretty straightforward fix. Call `getFaceInfo()` in `execute` if `_fe_problem.haveFV()`

lindsayad replied:
Oh never mind. I see that looping over face info for a 1-point quadrature rule on a coarse face doesn't make much sense

lindsayad replied:
However, I think that you could relax you error somewhat. If the `_current_elem` is more refined than its active neighbor, then you could proceed. But if you don't want to support that case right now, I'll accept that

lindsayad replied:
Yea my shower thoughts were that in implementations where you are making functor evaluations based on face-information, we could
1.  keep my suggested code changes above
2. Still call `getFaceInfo` in the derived class (don't worry about modifying `execute` farther down)
3. Return from `computeQpIntegral` early with 0, if `_qp != 0`

But the suggestion of actually having a base class that does loops over face information is much better IMO

GiudGiud committed:
Update framework/include/userobject/SideUserObject.h

Expand the elem/side loop to faceInfos to capture extra faces during adaptivity

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud committed:
Move selection of integration type to SideIntegralPostprocessor

Add more comments to face integration of functors, disallow high order quadratures for now, refs #18086

Delete InternalVolumetricFlowRate, it's the same now that we ve moved the interpolation abstraction to the functors

GiudGiud committed:
Rework volumetricFlowRate to follow new paradigm for postprocessor integration, refs #18086

GiudGiud committed:
Better documentation for the NS volumetric flow rate postprocessor, refs #18086

GiudGiud commented:
Should be ready again.

The change of the integration loop changes the base classes. I think it makes sense though, we may even have an enum on the integration type some day, not just qp or FI.

lindsayad commented on framework/src/postprocessors/SideIntegralFunctorPostprocessor.C:
```diff
@@ -0,0 +1,103 @@
+    has_elem = true;
+  else
+  {
+    if (fi->neighborPtr())
```
```suggestion
#ifndef NDEBUG
    if (fi->neighborPtr())
```

lindsayad commented on framework/src/postprocessors/SideIntegralFunctorPostprocessor.C:
```diff
@@ -0,0 +1,103 @@
+      mooseAssert(_prefactor.hasBlocks(_current_elem->subdomain_id()) ||
+                      _prefactor.hasBlocks(fi->neighborPtr()->subdomain_id()),
+                  "Prefactor should be defined on at least one side of the boundary");
+    }
```
```suggestion
    }
#endif
```

lindsayad commented on framework/src/postprocessors/SideIntegralPostprocessor.C:
```diff
@@ -53,7 +53,16 @@ Real
+    getFaceInfos();
+
+    for (auto & fi : _face_infos)
+      sum += _JxW[0] * _coord[0] * computeFaceInfoIntegral(fi);
```
```suggestion
      sum += fi->faceArea() * fi->faceCoord() * computeFaceInfoIntegral(fi);
```

lindsayad commented on framework/src/userobject/SideUserObject.C:
```diff
@@ -46,3 +46,49 @@ SideUserObject::SideUserObject(const InputParameters & parameters)
+  _face_infos.clear();
+
+  // Either the element or the (active) neighbor is a valid argument to get a face info
+  const Elem * equal_level_neighbor = _current_elem->neighbor_ptr(_current_side);
```
I realize now that this may not be an equal level neighbor, so let's change the name of this local variable. Maybe `side_neighbor`?

lindsayad commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -3232,7 +3230,7 @@ MooseMesh::buildFaceInfo() const
           (neighbor->active() && (neighbor->level() == elem->level()) &&
            (elem_id < neighbor->id())) ||
-          (neighbor->level() < elem->level()))
+          (neighbor->active() && neighbor->level() < elem->level()))
```
I don't think we should add this. If you want to add something, I would support adding a `mooseAssert` below. Something like
```
mooseAssert(neighbor->level() < elem->level() ? neighbor->active() : true, "If the neighbor is coarser than us, we expect that the neighbor must be active.");
```

GiudGiud replied:
works for me

lindsayad commented on modules/navier_stokes/src/fvbcs/INSFVOutletPressureBC.C:
```diff
@@ -9,22 +9,50 @@
   params += INSFVFullyDevelopedFlowBC::validParams();
+
+  // Value may be specified by a AD functor (typically a variable), a function or a postprocessor
+  params.addParam<FunctionName>("function", "The boundary pressure as a regular function");
```
Should we make this deprecated? Hmm I see that this could be important since the function returns Reals as opposed to the ADReal functor ... ?

GiudGiud replied:
yeah that's why I want to keep it.

GiudGiud commented on framework/include/postprocessors/SideIntegralFunctorPostprocessor.h:
```diff
@@ -0,0 +1,56 @@
+ * class and overriding computeQpIntegral().
+ */
+template <bool is_ad>
+class SideIntegralFunctorPostprocessorTempl : public SideIntegralPostprocessor
```
I say not in the name. In my mind we add the dirreferent integration methods as we need them

lindsayad replied:
ok that works

GiudGiud committed:
Apply suggestions from code review
Fixup some docstrings
Use face info centroid for coordinate for RZ integration

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented on framework/include/base/MooseFunctor.h:
```diff
@@ -327,6 +327,15 @@ class FunctorBase
+  /**
+   * Returns whether the functor is defined on this block
+   */
+  virtual bool hasBlocks(const SubdomainID & /* id */) const
```
the block restricted interface has a reference, and it s creating a "ambiguous" call where _var(block_id) could be one or the other. I d rather we stay away from this complexity and keep it consistent

lindsayad replied:
👍

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1026449/) on f9c2387 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20622/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20622/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a34443cae14ef47b63923eb4c3048458ce2d4921`

lindsayad commented on test/tests/postprocessors/side_integral/gold/side_integral_functor_refined_out.csv:
```diff
@@ -1,3 +1,3 @@
 time,ext_f1,ext_m1,ext_m2,ext_u,ext_v1,ext_v2,int_s1_f1,int_s1_m1,int_s1_m2,int_s1_u,int_s1_v1,int_s2_f1,int_s2_m1,int_s2_m2,int_s2_v2
 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
-1,202.0824437408,21.045455689911,132,5.4997921800047,28,0,9.8350801463159,9.8350801463159,84,3.4992744080908,22.5,19.670160292632,19.670160292632,56,45.15
+1,202.0824437408,21.045455689911,132,5.4997921800047,28,0,9.8350801463159,9.8350801463159,84,3.4992744080908,22.5,9.835080146316,9.835080146316,28,22.575
```
Did the changes in the last commit necessitate a re-gold?

GiudGiud replied:
the volume fix ? fi.faceArea()

lindsayad replied:
Makes me wonder what was going on with the quadrature rule. But these results look like "gold" results to you? E.g. are they better than before?

GiudGiud replied:
well the element isnt refined on the side it s looking at. I ll do the analytical calc

GiudGiud replied:
so m2 is the easiest one to integrate. it's value is 4 over a surface that is a u of length: 3, 1, 3
4 * (3 + 1 + 3) = 28, which is what we get

GiudGiud commented:
this is seg faulting on everything debug and fv right now, ill see if I can fix it

GiudGiud commented:
ok found the problem, new assert was dereferencing null ptr. Should be good now

moosebuild commented:
Job [Framework increased sparsity](https://civet.inl.gov/job/1027302/) on b4d0cd1 : invalidated by @lindsayad

GiudGiud commented:
@lindsayad your PR then this one? or the other way?

GiudGiud committed:
Adapt to changes in FV arguments, refs #20493

GiudGiud commented:
ok I rebased on next

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1027901/) on 4b79aa1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20622/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20622/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b7142812a21efa432f74506531784dfe3e8bf790`

GiudGiud commented:
conflicts were not so bad. I was a little cavalier with deleting InternalVolumetricFlowRate (it was inheriting from a different UO class, so i d have had to copy-paste all my changes to that one too). 

Though my experience so far has been that users are using VolumetricFlowRate everywhere, even internally, so we should just have a single object so they cant get it wrong

lindsayad commented on modules/navier_stokes/src/postprocessors/VolumetricFlowRate.C:
```diff
@@ -85,22 +85,37 @@ VolumetricFlowRate::computeFaceInfoIntegral([[maybe_unused]] const FaceInfo * fi
-                                              correct_skewness,
-                                              _current_elem->subdomain_id()});
+  // External faces for the advected quantity
+  if (!fi->neighborPtr() || !_adv_quant->hasBlocks(fi->neighborPtr()->subdomain_id()))
```
I like this!

GiudGiud commented:
pgh diff is because i rebased on next btw. It wasnt diffing before, nor should it be
